### PR TITLE
statistics: add lru cache implement for statsCacheInner

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

add lru cache implement for statsCacheInner.

The evict policy is desribed in this [document](https://pingcap.feishu.cn/docs/doccnM5l6gyLNvFKrRLKA0vCGwc#)

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/lru_cache.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+	// move index element
+	for idxID := range element.tblMemUsage.IndicesMemUsage {
+		s.lru.get(tblID, idxID)
+	}
+	return element.tbl, true
+}
+
+// Get implements statsCacheInner
+func (s *statsInnerCache) Get(tblID int64) (*statistics.Table, bool) {
+	s.Lock()
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply statistics/handle/lru_cache.go:85]: No, lock is necessary. `move element` is a write op.

File: statistics/handle/lru_cache.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+	needEvict, needMove bool) {
+	defer func() {
+		if needEvict {
+			c.evictIfNeeded()
+		}
+	}()
+	if c.capacity < itemMem.TrackingMemUsage() {
+		item.DropEvicted()
+		itemMem = item.MemoryUsage()
+	}
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply statistics/handle/lru_cache.go:312]: updated.

File: statistics/handle/lru_cache.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+		deletedIdx := make([]int64, 0)
+		for oldIdxID := range oldtbl.Indices {
+			_, exist := tbl.Indices[oldIdxID]
+			if !exist {
+				deletedIdx = append(deletedIdx, oldIdxID)
+			}
+		}
+		for idxID, index := range tbl.Indices {
+			idxMem := tblMemUsage.IndicesMemUsage[idxID]
+			if idxMem.TrackingMemUsage() < 1 {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply statistics/handle/lru_cache.go]: If idxMem.TrackingMemUsage() is larger than capactiy before putting, we will do `drop evict op` which may caused `idxMem.TrackingMemUsage() < 1`. As the above code is removed, so as these codes.

File: statistics/handle/lru_cache.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	sync.RWMutex
+	elements map[int64]*lruMapElement
+	// lru maintains item lru cache
+	lru *innerItemLruCache
+}
+
+func newStatsLruCache(c int64) *statsInnerCache {
+	s := &statsInnerCache{
+		elements: make(map[int64]*lruMapElement),
+		lru:      newInnerIndexLruCache(c),
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply statistics/handle/lru_cache.go]: I renamed `newInnerIndexLruCache` as newInnerLruCache` now.

File: statistics/handle/lru_cache.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+// GetByQuery implements statsCacheInner
+func (s *statsInnerCache) GetByQuery(tblID int64) (*statistics.Table, bool) {
+	s.Lock()
+	defer s.Unlock()
+	element, ok := s.elements[tblID]
+	if !ok {
+		return nil, false
+	}
+	// move index element
+	for idxID := range element.tblMemUsage.IndicesMemUsage {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply statistics/handle/lru_cache.go:77]: cuz we only need index id here.

File: statistics/handle/lru_cache.go [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+	element.tblMemUsage = element.tbl.MemoryUsage()
+}
+
+func (s *statsInnerCache) freshTableCost(tblID int64, element *lruMapElement) {
+	newTblMem := element.tbl.MemoryUsage()
+	for idxID, idx := range element.tbl.Indices {
+		s.lru.put(tblID, idxID, idx, newTblMem.IndicesMemUsage[idxID], true, false)
+	}
+	// tbl mem usage might be changed due to evict
+	element.tblMemUsage = element.tbl.MemoryUsage()
```
Comment (chrysan) [inline statistics/handle/lru_cache.go:270]: Is it duplicated with element.tbl.MemoryUsage() which is called in onEvict?

Comment (Yisaer) [reply statistics/handle/lru_cache.go]: right, it's unnecessary now, updated.

File: statistics/handle/lru_cache.go [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+	return &innerItemLruCache{
+		capacity: c,
+		cache:    list.New(),
+		elements: make(map[int64]map[int64]*list.Element, 0),
+	}
+}
+
+type lruCacheItem struct {
+	tblID         int64
+	id            int64
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache.go:58]: Do we just use `Index.ID` and `Column.ID` to set `lruCacheItem.id` here?

Comment (xuyifangreeneyes) [reply statistics/handle/lru_cache.go:58]: We will differentiate between index and column in the next pr.

Comment (Yisaer) [reply statistics/handle/lru_cache.go:58]: right

File: statistics/handle/lru_cache.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+	defer s.RUnlock()
+	return s.lru.trackingCost
+}
+
+func (s *statsInnerCache) totalCost() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	totalCost := int64(0)
+	for tblID, ele := range s.elements {
+		s.freshTableCost(tblID, ele)
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache.go:178]: Is it safe to call `freshTableCost` with `RLock`? `freshTableCost` call `innerItemLruCache.put` which does some writes.

Comment (Yisaer) [reply statistics/handle/lru_cache.go:178]: nice catch, updated.

File: statistics/handle/lru_cache.go [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+func (c *innerItemLruCache) get(tblID, id int64) (*lruCacheItem, bool) {
+	v, ok := c.elements[tblID]
+	if !ok {
+		return nil, false
+	}
+	ele, ok := v[id]
+	if !ok {
+		return nil, false
+	}
+	return ele.Value.(*lruCacheItem), true
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache.go:284]: Do we need to call `c.cache.MoveToFront(ele)` here?

Comment (Yisaer) [reply statistics/handle/lru_cache.go:285]: nice catch, updated.

File: statistics/handle/lru_cache.go [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+func (c *innerItemLruCache) copy() *innerItemLruCache {
+	newLRU := newInnerLruCache(c.capacity)
+	curr := c.cache.Back()
+	for curr != nil {
+		item := curr.Value.(*lruCacheItem)
+		newLRU.put(item.tblID, item.id, item.innerItem, item.innerMemUsage, false, false)
+		curr = curr.Prev()
+	}
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache.go:367]: Do we need to copy `c.onEvict` here?

File: statistics/handle/lru_cache_test.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func TestLRUPutTooBig(t *testing.T) {
+	lru := newStatsLruCache(1)
+	mockTable := newMockStatisticsTable(1, 1)
+	// put mockTable, the index should be evicted
+	lru.Put(int64(1), mockTable)
+	_, ok := lru.Get(int64(1))
+	require.True(t, ok)
+	require.Equal(t, lru.TotalCost(), indexMemoryUsage)
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache_test.go:217]:
```suggestion
	require.Equal(t, lru.TotalCost(), columnMemoryUsage)
```

File: statistics/handle/lru_cache_test.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+	t2ID := int64(1)
+	s.Put(t1ID, t1)
+	s.Put(t2ID, t2)
+	// assert t2 element should be front element
+	front := s.lru.cache.Front().Value.(*lruCacheItem)
+	require.Equal(t, t2ID, front.tblID)
+	// assert t1 element should be front element after GetByQuery
+	s.GetByQuery(t1ID)
+	front = s.lru.cache.Front().Value.(*lruCacheItem)
+	require.Equal(t, t2ID, front.tblID)
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache_test.go:257]:
```suggestion
	require.Equal(t, t1ID, front.tblID)
```

File: statistics/handle/lru_cache_test.go [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+	require.Equal(t, stats.Len(), 3)
+}
+
+func TestLRUMove(t *testing.T) {
+	capacity := int64(100)
+	s := newStatsLruCache(capacity)
+	t1 := newMockStatisticsTable(1, 1)
+	t1ID := int64(1)
+	t2 := newMockStatisticsTable(1, 1)
+	t2ID := int64(1)
```
Comment (xuyifangreeneyes) [inline statistics/handle/lru_cache_test.go:248]:
```suggestion
	t2ID := int64(2)
```

File: statistics/handle/lru_cache.go (hunk 10 above)
Comment (Yisaer) [reply statistics/handle/lru_cache.go:367]: no,  `onEvict` will be set in `newStatsLruCache`.

**Review by xuyifangreeneyes — APPROVED** [review]

File: statistics/handle/lru_cache.go (hunk 10 above)
Comment (Yisaer) [reply statistics/handle/lru_cache.go:368]: updated.

**Pull request merged**